### DISCUSSION TITLE
Building an OpenThread network in ns-3

### DISCUSSION THREAD

**Discussion by PradnyaYB** [openthread/openthread#9995] [General]
Hello all,

I am trying to build a OpenThread network of two to three router nodes in a simulation software called 'ns-3'. It doesn't have OpenThread models in it, but has lr-wpan model which resembles protocol stack, but has helper files for different router link formation and addressing scheme than OpenThread. May I get any idea on how I can proceed? I want to see if I can use OpenThread helper files in ns-3. (I want to analyze/experiment with thread routers hence the simulation.)

Thank you, any help/guidance is appreciated.

**Comment by jwhui** [comment]
@EskoDijk has been considering OTNS integration with ns-3. Maybe he can share some thoughts.

**Comment by EskoDijk** [comment]
Sure! I was sideways involved in a project to integrate NS-3 with OpenThread; based on an open-source framework called Whitefield. Experimental results are available here: https://github.com/Cascoda/whitefield/tree/ot-integration 

However I found this difficult to use for practical tasks (GUI wasn't working ; bugs etc.) and details of radio timing (e.g. CCA) was not investigated in full detail here. Meanwhile OTNS2 was developed during 2022-2024 as a more practical tool for OpenThread simulation that simulates more accurate radio behavior. It's hosted at https://github.com/EskoDijk/ot-ns/ -- soon to be merged back into the OpenThread project I hope!  It uses under the hood some radio models that NS-3 and the lr-wpan PHY can also offer.

Of course doing a real integration with NS-3 is still on the wish-list. This would enable useful things like mixing OpenThread nodes with other node types like simulated Wi-Fi transmissions, better usefulness for academic research and papers, maybe including 3D obstacles like floors/walls (I assume NS-3 can do this?) etc.

When using the lr-wpan model of NS-3 with OpenThread you have to be careful: lr-wpan MAC doesn't actually simulate the Thread MAC behavior, but rather a generic 802.15.4 MAC which is not how Thread will behave.  Only the lr-wpan PHY portion would be useful and re-usable to simulate the frame error probability and interference of 802.15.4 frames; which is a similar thing as done in OTNS2 now.  NS-3 can be linked to OpenThread nodes for sure but there's currently no maintained code available that does this.

**Reply by PradnyaYB** [reply]
@jwhui Thank you for connecting @EskoDijk in this Q&A thread.
@EskoDijk Thank you very much for this information and helpful links.
